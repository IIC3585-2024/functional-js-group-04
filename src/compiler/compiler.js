import { readFileSync, writeFileSync } from 'fs';
import { join, dirname } from 'path';
import _tokenize from "../tokenizer/tokenizer.js";
import _parse from "../parser/parser.js";

function compiler() {
    let _output = null;

    function tokenize() {
        _output = _tokenize(_output);
        return this
    }

    function parse() {
        _output = _parse(_output);
        return this;
    }

    function compile() {
        this.tokenize().parse();
        return this;
    }

    function output() {
        return _output;
    }

    function setSource(source) {
        _output = source;
        return this;
    }

    function write(out_path) {
        try {
            if (typeof _output !== 'string') {
                writeFileSync(out_path, JSON.stringify(_output, null, 2));
            } else {
                writeFileSync(out_path, _output);
            }    
        } catch (error) {
            console.error(`Error writing file to disk: ${error}`);
        }
        return this;
    }

    function style(out_path, st_path) {
        try {
            const styleContent = readFileSync(join('src', 'assets', st_path), 'utf-8');
            const cssFilePath = join(dirname(out_path), 'style.css');
            writeFileSync(cssFilePath, styleContent);
        } catch (error) {
            console.error(`Error writing style file to disk: ${error}`);
        }
        return this;
    }

    function read(in_path) {
        try {
            _output = readFileSync(in_path, 'utf8');
        } catch (error) {
            console.error(`Error reading file from disk: ${error}`);
        }
        return this;
    }

    return {
        tokenize,
        parse,
        compile,
        setSource,
        output,
        read,
        write,
        style,
    }
}

export default compiler;