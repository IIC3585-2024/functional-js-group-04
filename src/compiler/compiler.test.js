import compiler from "./compiler.js"


describe("#compiler", () => {
    test("is a function", () => {
        expect(typeof compiler).toBe("function");
    });

    test("complete chain of methods", () => {
        const input = `# Hello World`
        const output = compiler().setSource(input).tokenize().parse().output();
        expect(output).toMatchSnapshot();
    });

    test(".compile().setSource().output() returns a string", () => {
        const input = `# Hello World`
        const output = compiler().setSource(input).compile().output();
        expect(output).toMatchSnapshot();
    });
})
