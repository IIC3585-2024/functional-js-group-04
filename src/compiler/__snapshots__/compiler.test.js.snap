// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#compiler .compile().setSource().output() returns a string 1`] = `"<!DOCTYPE html><style>null</style><html><head><title>Markdown</title></head><body><h1>Hello World</h1></body></html>"`;

exports[`#compiler complete chain of methods 1`] = `"<!DOCTYPE html><style>null</style><html><head><title>Markdown</title></head><body><h1>Hello World</h1></body></html>"`;
